### PR TITLE
add meta data to contract to support shard contract

### PR DESCRIPTION
**Pull request by qiluge**
1. don't set meta data while **deploy** contract;
2. when **migrating** contract, update meta data only while it is not nil;
3. if meta data is nil, the contract can only be invoked at self ledger;
4. InitMetaData only can be invoked one time, and only can be invoked at self ledger;
5. if meta data is not nil, check the shardId whether contract can be invoked at current shard;
6. the new type 'MetaData' transaction only can be invoked at self ledger and require the contract has already initialized meta data.

### CODE REVIEW

File: core/payload/metadata_code.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package payload
```
Comment (Honglei-Cong) [inline core/payload/metadata_code.go:19]: add license header.

File: core/types/mutable_transaction.go [hunk 2]
```diff
@@ -118,15 +118,9 @@ func (tx *MutableTransaction) serializeUnsigned(sink *common.ZeroCopySink) error
...
-			return err
-		}
+		pl.Serialization(sink)
 	case *payload.InvokeCode:
-		err := pl.Serialization(sink)
-		if err != nil {
-			return err
-		}
+		pl.Serialization(sink)
 	default:
```
Comment (Honglei-Cong) [inline core/types/mutable_transaction.go:126]: need support MetadataType here?

File: core/store/ledgerstore/tx_handler.go [hunk 3]
```diff
@@ -122,6 +122,85 @@ func (self *StateStore) HandleDeployTransaction(store store.LedgerStore, overlay
...
+	defer func() {
+		if err != nil {
+			if err := costInvalidGas(tx.Payer, tx.GasLimit*tx.GasPrice, cfg, overlay, store, notify, shardID); err != nil {
+				log.Errorf("HandleChangeMetadataTransaction: costInvalidGas failed, err: %s", err)
+			} else {
+				cache.Commit()
+			}
+		}
+	}()
+	if tx.Version != common.VERSION_SUPPORT_SHARD {
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/tx_handler.go]: tx.Version >= VERSION_SUPPORT_SHARD ?

File: core/store/ledgerstore/tx_handler.go [hunk 4]
```diff
@@ -122,6 +122,85 @@ func (self *StateStore) HandleDeployTransaction(store store.LedgerStore, overlay
...
+	}
+	cfg := &smartcontract.Config{
+		ShardID:   shardID,
+		Time:      block.Header.Timestamp,
+		Height:    block.Header.Height,
+		Tx:        tx,
+		BlockHash: block.Hash(),
+	}
+	defer func() {
+		if err != nil {
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/tx_handler.go:140]: It seems difficult to handle all cases correctly in defer().  suggest not defer().

File: smartcontract/service/neovm/contract.go [hunk 5]
```diff
@@ -46,6 +47,30 @@ func ContractCreate(service *NeoVmService, engine *vm.ExecutionEngine) error {
...
+		return errors.NewDetailErr(err, errors.ErrNoCode, "[InitMetaData] meta data has already initialized")
+	}
+	newMeta, err := getMetaData(engine)
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, fmt.Sprintf("[InitMetaData] invalid param: %s", err))
+	}
+	if !checkInitMeta(service, newMeta) {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "[InitMetaData] meta data should contain owner")
+	}
+	newMeta.OntVersion = common.VERSION_SUPPORT_SHARD
```
Comment (Honglei-Cong) [inline smartcontract/service/neovm/contract.go]: set to CURRENT_VERSION

File: smartcontract/storage/cachedb.go [hunk 6]
```diff
@@ -117,24 +117,47 @@ func (self *CacheDB) GetContract(addr comm.Address) (*payload.DeployCode, error)
...
 }
 
+func (self *CacheDB) GetMetaData(addr comm.Address) (*payload.MetaDataCode, error) {
+	value, err := self.get(common.ST_CONTRACT_META_DATA, addr[:])
+	if err != nil {
+		return nil, err
+	}
+
+	if len(value) == 0 {
+		return nil, nil
```
Comment (Honglei-Cong) [inline smartcontract/storage/cachedb.go:141]: return default-metadata?

Comment (qiluge) [reply smartcontract/storage/cachedb.go:141]: use meta data == nil to judge is meta data initialized

File: core/store/ledgerstore/tx_handler.go (hunk 4 above)
Comment (qiluge) [reply core/store/ledgerstore/tx_handler.go:140]: all failed handle should be charged at defer func

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**